h/static/scripts/annotator/anchoring/pdf.coffee

```diff
@@ -0,0 +1,224 @@
+seek = require('dom-seek')
+
+Annotator = require('annotator')
+xpathRange = Annotator.Range
+
+html = require('./html')
+{TextPositionAnchor, TextQuoteAnchor} = require('./types')
+
+
+getSiblingIndex = (node) ->
+  siblings = Array.prototype.slice.call(node.parentNode.childNodes)
+  return siblings.indexOf(node)
+
+
+getNodeTextLayer = (node) ->
+  until node.classList?.contains('page')
+    node = node.parentNode
+  return node.getElementsByClassName('textLayer')[0]
+
+
+getPage = (pageIndex) ->
+  return PDFViewerApplication.pdfViewer.pages[pageIndex]
+
+
+getPageTextContent = (pageIndex, cache) ->
+  if cache[pageIndex]?
+    return Promise.resolve(cache[pageIndex])
+  else
+    joinItems = ({items}) ->
+      nonEmpty = (item.str for item in items when /\S/.test(item.str))
+      textContent = nonEmpty.join('')
+      cache[pageIndex] = textContent
+      return textContent
+
+    return PDFViewerApplication.pdfViewer.getPageTextContent(pageIndex)
+    .then(joinItems)
+
+getPageOffset = (pageIndex, cache = {}) ->
+  index = -1
+
+  next = (offset) ->
+    if ++index is pageIndex
+      return Promise.resolve(offset)
+
+    return getPageTextContent(index, cache)
+    .then((textContent) -> next(offset + textContent.length))
+
+  return next(0)
+
+
+
+findPage = (offset, cache = {}) ->
+  index = 0
+  total = 0
+
+  count = (textContent) ->
+    if total + textContent.length >= offset
+      offset = total
+      return Promise.resolve({index, offset, textContent})
+    else
+      index++
+      total += textContent.length
+      return getPageTextContent(index, cache).then(count)
+
+  return getPageTextContent(0, cache).then(count)
+
+
+###*
+# Anchor a set of selectors.
+#
+# This function converts a set of selectors into a document range using.
+# It encapsulates the core anchoring algorithm, using the selectors alone or
+# in combination to establish the best anchor within the document.
+#
+# :param Element root: The root element of the anchoring context.
+# :param Array selectors: The selectors to try.
+# :param Object options: Options to pass to the anchor implementations.
+# :return: A Promise that resolves to a Range on success.
+# :rtype: Promise
+####
+exports.anchor = (root, selectors, options = {}) ->
```

Comment:
I think this needs some cleaning up. The placeholder stuff is confusing and could use comments. The `anchorByPosition` function could be broken out for readability and given parameters rather than closing over the local scope.